pkg/snmp/traps/server_test.go

```diff
@@ -0,0 +1,242 @@
+		close(packets)
+		t.Errorf("Trap not received")
+		return nil
+	}
+}
+
+func assertV2c(t *testing.T, p *gosnmp.SnmpPacket, config TrapListenerConfig) {
+	require.Equal(t, gosnmp.Version2c, p.Version)
+	require.Equal(t, config.Community, p.Community)
+}
+
+// func assertV3(t *testing.T, p *gosnmp.SnmpPacket, config TrapListenerConfig) {
+// 	require.Equal(t, gosnmp.Version3, p.Version)
+// }
+
+func assertVariables(t *testing.T, p *gosnmp.SnmpPacket) {
+	assert.Equal(t, 4, len(p.Variables))
+	uptime := p.Variables[0]
+	assert.Equal(t, ".1.3.6.1.2.1.1.3.0", uptime.Name)
+	assert.Equal(t, gosnmp.TimeTicks, uptime.Type)
+
+	snmptrapOID := p.Variables[1]
+	assert.Equal(t, ".1.3.6.1.6.3.1.1.4.1", snmptrapOID.Name)
+	assert.Equal(t, gosnmp.OctetString, snmptrapOID.Type)
+	assert.Equal(t, "1.3.6.1.4.1.8072.2.3.0.1", string(snmptrapOID.Value.([]byte)))
+
+	heartBeatRate := p.Variables[2]
+	assert.Equal(t, ".1.3.6.1.4.1.8072.2.3.2.1", heartBeatRate.Name)
+	assert.Equal(t, gosnmp.Integer, heartBeatRate.Type)
+	assert.Equal(t, 1024, heartBeatRate.Value.(int))
+
+	heartBeatName := p.Variables[3]
+	assert.Equal(t, ".1.3.6.1.4.1.8072.2.3.2.2", heartBeatName.Name)
+	assert.Equal(t, gosnmp.OctetString, heartBeatName.Type)
+	assert.Equal(t, "test", string(heartBeatName.Value.([]byte)))
+}
+
+func TestTraps(t *testing.T) {
+	t.Run("empty", func(t *testing.T) {
+		s, err := NewTrapServer()
+		require.NoError(t, err)
+		assert.NotNil(t, s)
+		defer s.Stop()
+		assert.True(t, s.Started)
+	})
+
+	t.Run("v2c-single", func(t *testing.T) {
+		port, err := getAvailableUDPPort()
+		require.NoError(t, err)
+
+		config := TrapListenerConfig{
+			Port:      port,
+			Community: "public",
+		}
+
+		configure(t, fmt.Sprintf(`
+snmp_traps_listeners:
+  - port: %d
+    community: %s
+`, config.Port, config.Community))
+
+		s, err := NewTrapServer()
+		require.NoError(t, err)
+		require.NotNil(t, s)
+		defer s.Stop()
+		require.True(t, s.Started)
+		require.Equal(t, s.NumListeners(), 1)
+
+		p := sendTestTrap(t, s, config)
+		assertV2c(t, p, config)
+		assertVariables(t, p)
+	})
+
+	t.Run("v3-single", func(t *testing.T) {
+		port, err := getAvailableUDPPort()
+		require.NoError(t, err)
+
+		config := TrapListenerConfig{
+			Port:         port,
+			User:         "doggo",
+			AuthProtocol: "MD5",
+			AuthKey:      "doggopass",
+			PrivProtocol: "DES",
+			PrivKey:      "doggokey",
+		}
+
+		configure(t, fmt.Sprintf(`
+snmp_traps_listeners:
+  - port: %d
+    user: %s
+    auth_protocol: %s
+    auth_key: %s
+    priv_protocol: %s
+    priv_key: %s
+`, config.Port, config.User, config.AuthProtocol, config.AuthKey, config.PrivProtocol, config.PrivKey))
+
+		s, err := NewTrapServer()
+		require.NoError(t, err)
+		require.NotNil(t, s)
+		defer s.Stop()
+		require.True(t, s.Started)
+		require.Equal(t, s.NumListeners(), 1)
+
+		// FIXME: this triggers a panic during trap unmarshalling
+		// p := sendTestTrap(t, s, config)
+		// assertV3(t, p, config)
+		// assertVariables(t, p)
```

Comment:
Might be a manifestation of _"[NOTE: the trap code is currently unreliable when working with snmpv3](https://godoc.org/github.com/soniah/gosnmp#GoSNMP.SendTrap)"_.

Full stack trace:

```console
--- Running unit tests:
=== RUN   TestTraps
=== RUN   TestTraps/empty
=== RUN   TestTraps/v2c-single
=== RUN   TestTraps/v3-single
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x1214d96]

goroutine 32 [running]:
github.com/soniah/gosnmp.(*UsmSecurityParameters).unmarshal(0xc0001ea340, 0x4, 0xc0001c3000, 0x40, 0x1000, 0x21, 0xc0001a85c8, 0x3, 0x0)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/v3_usm.go:879 +0xeb6
github.com/soniah/gosnmp.(*GoSNMP).unmarshalV3Header(0xc0001d4360, 0xc0001c3000, 0x40, 0x1000, 0x18, 0xc0001049a0, 0xc0001a8598, 0x3, 0x0)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/v3.go:407 +0x4cd
github.com/soniah/gosnmp.(*GoSNMP).unmarshalHeader(0xc0001d4360, 0xc0001c3000, 0x40, 0x1000, 0xc0001049a0, 0xc00006ee50, 0x11c892d, 0xc0000b0220)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/marshal.go:820 +0x5b4
github.com/soniah/gosnmp.(*GoSNMP).UnmarshalTrap(0xc0001d4360, 0xc0001c3000, 0x40, 0x1000, 0x40)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:296 +0xa0
github.com/soniah/gosnmp.(*TrapListener).listenUDP(0xc0000d4f00, 0xc0001a8390, 0xf, 0x0, 0x0)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:174 +0x2cd
github.com/soniah/gosnmp.(*TrapListener).Listen(0xc0000d4f00, 0xc0001a8390, 0xf, 0x1, 0x1)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:277 +0x1c5
github.com/DataDog/datadog-agent/pkg/snmp/traps.(*TrapListener).Listen(0xc000179b40)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/pkg/snmp/traps/listener.go:53 +0xcd
created by github.com/DataDog/datadog-agent/pkg/snmp/traps.(*TrapServer).start
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/pkg/snmp/traps/server.go:71 +0x17f
FAIL    github.com/DataDog/datadog-agent/pkg/snmp/traps 0.344s
FAIL
```